File: BookLand_FE/src/types/Event.ts
import type { User } from './User';
import type { EventImage } from './EventImage';
import type { EventTarget } from './EventTarget';
import type { EventRule } from './EventRule';
import type { EventAction } from './EventAction';
import type { EventLog } from './EventLog';
import type { EventType } from './EventType';

export const EventStatus = {
    DRAFT: 'DRAFT',
    ACTIVE: 'ACTIVE',
    PAUSED: 'PAUSED',
    EXPIRED: 'EXPIRED',
    DISABLED: 'DISABLED'
} as const;

export type EventStatus = (typeof EventStatus)[keyof typeof EventStatus];

export interface Event {
    id: number;
    name: string;
    description?: string;
    type: EventType;
    startTime: string;
    endTime: string;
    status: EventStatus;
    priority: number;
    createdBy?: User;
    createdAt?: string;
    images?: EventImage[];
    targets?: EventTarget[];
    rules?: EventRule[];
    actions?: EventAction[];
    logs?: EventLog[];
}

export interface EventRequest {
    name: string;
    description?: string;
    type: EventType;
    startTime: string;
    endTime: string;
    status: EventStatus;
    priority: number;
    images?: Omit<EventImage, 'id' | 'event'>[];
    targets?: Omit<EventTarget, 'id' | 'event'>[];
    rules?: Omit<EventRule, 'id' | 'event'>[];
    actions?: Omit<EventAction, 'id' | 'event'>[];
}

export interface EventPayload {
    name: string;
    description?: string;
    type: EventType;
    startTime: string;
    endTime: string;
    status: EventStatus;
    priority: number;
    createdById: number;
    mainImageUrls: string[];
    subImageUrls: string[];
    targets?: Omit<EventTarget, 'id' | 'event'>[];
    rules?: Omit<EventRule, 'id' | 'event'>[];
    actions?: Omit<EventAction, 'id' | 'event'>[];
}

File: services/event-service/src/main/java/com/bookland/event/entity/Event.java
package com.bookland.event.entity;

import com.bookland.event.enums.EventType;
import lombok.*;
import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "event")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Event {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private EventType type;

    @Column(nullable = false)
    private LocalDateTime startTime;

    @Column(nullable = false)
    private LocalDateTime endTime;

    @Enumerated(EnumType.STRING)
    @Builder.Default
    private EventStatus status = EventStatus.DRAFT;

    @Builder.Default
    private Integer priority = 0;

    @Column(name = "createdBy")
    private String createdBy;

    @Column(updatable = false)
    private LocalDateTime createdAt;

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<EventImage> images = new HashSet<>();

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<EventTarget> targets = new HashSet<>();

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<EventRule> rules = new HashSet<>();

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private Set<EventAction> actions = new HashSet<>();

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL)
    @Builder.Default
    private Set<EventLog> logs = new HashSet<>();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }

    @Transient
    public boolean isActive() {
        LocalDateTime now = LocalDateTime.now();
        return status == EventStatus.ACTIVE
                && now.isAfter(startTime)
                && now.isBefore(endTime);
    }

    public enum EventStatus {
        DRAFT, ACTIVE, PAUSED, EXPIRED, DISABLED
    }
}
